File: src/shared/telemetry/go/telemetry.go
// Copyright 2026 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");

// Package telemetry configures OTLP tracing and defines the cross-language
// conventions used for NATS spans in the demo.
package telemetry

import (
	"context"
	"fmt"
	"os"
	"strings"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
)

const (
	// CorrelationIDKey is deliberately separate from the OpenTelemetry trace
	// ID. Correlation IDs are stable business/request identifiers and may not
	// satisfy the W3C trace-ID format.
	CorrelationIDKey = "correlation.id"
	instrumentation  = "online-boutique/messaging"
)

// Init installs W3C propagation and, when ENABLE_TRACING=1, an OTLP/gRPC
// exporter. The returned shutdown function is always safe to call.
func Init(ctx context.Context, serviceName string) (func(context.Context) error, error) {
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{}, propagation.Baggage{}))
	if os.Getenv("ENABLE_TRACING") != "1" {
		return func(context.Context) error { return nil }, nil
	}
	endpoint := os.Getenv("COLLECTOR_SERVICE_ADDR")
	if endpoint == "" {
		endpoint = strings.TrimPrefix(os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT"), "http://")
		endpoint = strings.TrimPrefix(endpoint, "https://")
	}
	if endpoint == "" {
		return nil, fmt.Errorf("COLLECTOR_SERVICE_ADDR or OTEL_EXPORTER_OTLP_ENDPOINT is required when tracing is enabled")
	}
	exporter, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpoint(endpoint), otlptracegrpc.WithInsecure())
	if err != nil {
		return nil, fmt.Errorf("create OTLP trace exporter: %w", err)
	}
	res, err := resource.Merge(resource.Default(), resource.NewWithAttributes(
		"", attribute.String("service.name", serviceName)))
	if err != nil {
		return nil, fmt.Errorf("create OpenTelemetry resource: %w", err)
	}
	provider := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sdktrace.ParentBased(sdktrace.AlwaysSample())),
	)
	otel.SetTracerProvider(provider)
	return provider.Shutdown, nil
}

// SetCorrelationID adds the request/business correlation ID to the current
// span. Empty values are represented consistently with the logs.
func SetCorrelationID(ctx context.Context, correlationID string) {
	trace.SpanFromContext(ctx).SetAttributes(
		attribute.String(CorrelationIDKey, normalized(correlationID)))
}

// StartConsumerSpan extracts W3C context from a message envelope and starts a
// NATS consumer span carrying the same correlation ID used by logs.
func StartConsumerSpan(ctx context.Context, subject, kind, messageID, correlationID, traceparent, tracestate string) (context.Context, trace.Span) {
	carrier := propagation.MapCarrier{}
	if traceparent != "" {
		carrier.Set("traceparent", traceparent)
	}
	if tracestate != "" {
		carrier.Set("tracestate", tracestate)
	}
	ctx = otel.GetTextMapPropagator().Extract(ctx, carrier)
	return otel.Tracer(instrumentation).Start(ctx, subject+" process",
		trace.WithSpanKind(trace.SpanKindConsumer),
		trace.WithAttributes(messageAttributes(subject, kind, messageID, correlationID)...))
}

// StartProducerSpan starts a NATS producer span for an outgoing envelope.
func StartProducerSpan(ctx context.Context, subject, kind, messageID, correlationID string) (context.Context, trace.Span) {
	return otel.Tracer(instrumentation).Start(ctx, subject+" publish",
		trace.WithSpanKind(trace.SpanKindProducer),
		trace.WithAttributes(messageAttributes(subject, kind, messageID, correlationID)...))
}

// Inject writes the current W3C context into an envelope's propagation fields.
func Inject(ctx context.Context, traceparent, tracestate *string) {
	carrier := propagation.MapCarrier{}
	otel.GetTextMapPropagator().Inject(ctx, carrier)
	*traceparent = carrier.Get("traceparent")
	*tracestate = carrier.Get("tracestate")
}

// RecordError marks a span failed without exposing error details as attributes.
func RecordError(span trace.Span, err error) {
	if err == nil {
		return
	}
	span.RecordError(err)
	span.SetStatus(codes.Error, "operation failed")
}

func messageAttributes(subject, kind, messageID, correlationID string) []attribute.KeyValue {
	return []attribute.KeyValue{
		attribute.String("messaging.system", "nats"),
		attribute.String("messaging.destination.name", subject),
		attribute.String("messaging.operation.type", kind),
		attribute.String("messaging.message.id", normalized(messageID)),
		attribute.String(CorrelationIDKey, normalized(correlationID)),
	}
}

func normalized(value string) string {
	if value == "" {
		return "unknown"
	}
	return value
}

File: src/adservice/src/main/java/hipstershop/Telemetry.java
/* Copyright 2026 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License"); */

package hipstershop;

import boutique.common.v1.MessageEnvelope;
import io.opentelemetry.api.OpenTelemetry;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanKind;
import io.opentelemetry.api.trace.StatusCode;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.context.Context;
import io.opentelemetry.context.Scope;
import io.opentelemetry.context.propagation.TextMapGetter;
import io.opentelemetry.context.propagation.TextMapSetter;
import io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdk;
import java.util.List;
import org.apache.logging.log4j.ThreadContext;

final class Telemetry {
  private static boolean enabled;
  private static final TextMapGetter<MessageEnvelope> GETTER =
      new TextMapGetter<>() {
        @Override
        public Iterable<String> keys(MessageEnvelope carrier) {
          return List.of("traceparent", "tracestate");
        }

        @Override
        public String get(MessageEnvelope carrier, String key) {
          if (carrier == null) {
            return null;
          }
          return switch (key) {
            case "traceparent" -> carrier.getTraceparent();
            case "tracestate" -> carrier.getTracestate();
            default -> null;
          };
        }
      };
  private static final TextMapSetter<MessageEnvelope.Builder> SETTER =
      (carrier, key, value) -> {
        if (carrier == null) {
          return;
        }
        if ("traceparent".equals(key)) {
          carrier.setTraceparent(value);
        } else if ("tracestate".equals(key)) {
          carrier.setTracestate(value);
        }
      };

  private static OpenTelemetry openTelemetry = OpenTelemetry.noop();
  private static Tracer tracer = openTelemetry.getTracer("online-boutique/messaging");

  private Telemetry() {}

  static void initialize() {
    if (!"1".equals(System.getenv("ENABLE_TRACING"))) {
      return;
    }
    enabled = true;
    openTelemetry = AutoConfiguredOpenTelemetrySdk.initialize().getOpenTelemetrySdk();
    tracer = openTelemetry.getTracer("online-boutique/messaging");
  }

  static MessageSpan consumer(MessageEnvelope envelope, String subject, String kind) {
    if (!enabled) {
      return MessageSpan.noop();
    }
    Context parent =
        openTelemetry.getPropagators().getTextMapPropagator().extract(Context.root(), envelope, GETTER);
    String correlationId = normalized(envelope == null ? "" : envelope.getCorrelationId());
    String messageId = normalized(envelope == null ? "" : envelope.getMessageId());
    Span span =
        tracer
            .spanBuilder(subject + " process")
            .setParent(parent)
            .setSpanKind(SpanKind.CONSUMER)
            .setAttribute("messaging.system", "nats")
            .setAttribute("messaging.destination.name", subject)
            .setAttribute("messaging.operation.type", kind)
            .setAttribute("messaging.message.id", messageId)
            .setAttribute("correlation.id", correlationId)
            .startSpan();
    return new MessageSpan(span);
  }

  static void inject(MessageEnvelope.Builder envelope) {
    if (!enabled) {
      return;
    }
    openTelemetry
        .getPropagators()
        .getTextMapPropagator()
        .inject(Context.current(), envelope, SETTER);
  }

  private static String normalized(String value) {
    return value == null || value.isBlank() ? "unknown" : value;
  }

  static final class MessageSpan implements AutoCloseable {
    private static final MessageSpan NOOP = new MessageSpan();
    private final Span span;
    private final Scope scope;

    private MessageSpan() {
      this.span = null;
      this.scope = null;
    }

    static MessageSpan noop() {
      return NOOP;
    }

    private MessageSpan(Span span) {
      this.span = span;
      this.scope = span.makeCurrent();
      if (span.getSpanContext().isValid()) {
        ThreadContext.put("traceId", span.getSpanContext().getTraceId());
        ThreadContext.put("spanId", span.getSpanContext().getSpanId());
        ThreadContext.put("traceSampled", Boolean.toString(span.getSpanContext().isSampled()));
      }
    }

    void recordError(Throwable error) {
      if (span == null) {
        return;
      }
      span.recordException(error);
      span.setStatus(StatusCode.ERROR, "operation failed");
    }

    @Override
    public void close() {
      if (span == null) {
        return;
      }
      scope.close();
      span.end();
      ThreadContext.remove("traceId");
      ThreadContext.remove("spanId");
      ThreadContext.remove("traceSampled");
    }
  }
}
